### ISSUE TITLE
Can not search `what` `why` or `how` in unencrypted chat

### ISSUE THREAD

**Issue by hoangdat** [linagora/twake-on-matrix#1786]
- [ ] global search
- [ ] chat search

**Comment by Te-Z** [comment]
IMO it is a back end problem: I found out that some words works and other don't. 
What I tested for example:
- search yes works, but ye does not
- yo work but you does not



![Image](https://github.com/linagora/twake-on-matrix/assets/31937920/ff6230fa-00a5-406e-8ef4-c670d109e6b1)
![Image](https://github.com/linagora/twake-on-matrix/assets/31937920/a869194a-8b61-4fc8-8905-bb031e108531)

**Comment by Te-Z** [comment]
@hoangdat FYI message sent to Xavier about this problem. Waiting for any feedback from him or you

**Comment by guimard** [comment]
Hi, this is due to the minimal size of requests. What could be the good minimal size ?

**Comment by chibenwa** [comment]
>  IMO it is a back end problem: I found out that some words works and other don't.

https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-stop-tokenfilter.html

stop words are typically excluded from search as they do not add relevance but costs a lot to manage. Thats how full text search works. full text search is not exact search.

Based on the requirements, the underlying mapping may be tunned, eg adapt language, custom filters etc... Carefull all of this can be complex functionnally and have performance implications.

I can help doing a mapping review.

**Comment by chibenwa** [comment]
Though can you please provide the mapping and exact Opensearch queries?

**Comment by Te-Z** [comment]
Hi @chibenwa is this question for me or @guimard ?

**Comment by hoangdat** [comment]
Open Search right now is not deployed. @guimard please try to deploy open search in TWP platforms to verify the problem.

Thanks

**Comment by tk-nguyen** [comment]
> > IMO it is a back end problem: I found out that some words works and other don't.
> 
> https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-stop-tokenfilter.html
> 
> stop words are typically excluded from search as they do not add relevance but costs a lot to manage. Thats how full text search works. full text search is not exact search.
> 
> Based on the requirements, the underlying mapping may be tunned, eg adapt language, custom filters etc... Carefull all of this can be complex functionnally and have performance implications.
> 
> I can help doing a mapping review.

This is the main point, though Synapse is using Postgres for its full text search, not opensearch/elasticsearch.

This issue is reported: https://github.com/matrix-org/synapse/issues/10559

Someone has made a breakdown of how Synapse search works: https://github.com/bkil/wiki/blob/master/en/dev/matrix-full-text-search.md

**Comment by dab246** [comment]
Also can not search `can` or `you` in unencrypted chat.

Env: `chat.stg.lin-saas.com` - `Chrome` & `Edge (Macbook)`


https://github.com/user-attachments/assets/ff6b077f-480a-437e-8570-5ae55e8e2190

**Comment by DianaPotokina** [comment]
@hoangdat @pm-McFly please fix that. Merci

**Comment by pm-McFly** [comment]
# TL:DR ;

This behavior, of not being able to search for certain words, is neither consistent neither an error. It mostly depends on the database engine used to store and search messages, but also to the actual search content.

If looking for only "why" doesn't rise up any results, try a more complete/meaningful search: "why is the color"

![Image](https://github.com/user-attachments/assets/ee89f5b1-d132-4f1c-b6ff-6cad58252222)

You'll then see that "why" is correctly colored as a match to the search.

# Details

> Someone has made a breakdown of how Synapse search works: https://github.com/bkil/wiki/blob/master/en/dev/matrix-full-text-search.md

As the link explain, some words are defined as "stop words" - meaning they are not meaningful enough to be used as searched items, or searching them would imply returning thousands of matches.

For example returning all occurrences of "the" would result in an absolute mess of partial and total match as "the" is widely used in English. Analog, "interrogative words" (why, what, etc.) could match almost all received messages holding a question in it which would be completely irrelevant.

![Image](https://github.com/user-attachments/assets/47b0fcbc-9b47-4f55-9642-fd0fa1bf56ce)

But, a searched query with sufficient "meaning" - including stop words and "normal" words - would result in the possibility of matching these stop words as they're now part of a plausible search:

![Image](https://github.com/user-attachments/assets/6765c2e9-f520-4977-aa13-aa987245b4fa)

It exists many stop words lists, one of the most popular is: https://github.com/stopwords-iso (for English text: https://github.com/stopwords-iso/stopwords-en/blob/master/stopwords-en.txt)

As we are relying mostly on Postgres for our deployed services this list could be in use: https://github.com/postgres/postgres/blob/master/src/backend/snowball/stopwords/english.stop

PSQL documentation about stop words: https://www.postgresql.org/docs/current/textsearch-dictionaries.html#TEXTSEARCH-STOPWORDS

If I am trying the same with a local deployed instance using sqlite, I got:

![Image](https://github.com/user-attachments/assets/c47a12b7-4dff-49b1-805e-2001dc9f5709)

This behavior is then normal because sqlite doesn't come with a built-in tokenizer that include such "stop words" and thus allows more matches (https://stackoverflow.com/questions/6314001/sqlite3-fts3-and-stop-words)
